src/charm.py

```diff
@@ -133,32 +134,40 @@ def _configure(self, _):
         # Restart prometheus only if command line arguments have changed,
         # otherwise just reload its configuration.
         if current_services == new_layer.services:
+            # No change in layer; reload config to make sure it is valid
             external_url = urlparse(self._external_url)
-
             prometheus_server = Prometheus(web_route_prefix=external_url.path)
-
             reloaded = prometheus_server.reload_configuration()
             if not reloaded:
                 logger.error("Prometheus failed to reload the configuration")
                 self.unit.status = BlockedStatus(CORRUPT_PROMETHEUS_CONFIG_MESSAGE)
                 return
+
             logger.info("Prometheus configuration reloaded")
+
         else:
+            # Layer changed - replan.
             container.add_layer(self._name, new_layer, combine=True)
             try:
+                # If a config is invalid then prometheus would exist immediately.
+                # This would be caught by pebble (default timeout is 30 sec) and a ChangeError
+                # would be raised.
                 container.replan()
                 logger.info("Prometheus (re)started")
-            except Exception as e:
-                logger.error(f"Failed to replan; pebble plan: {container.get_plan().to_dict()}")
-                raise e
-
-        if (
-            isinstance(self.unit.status, BlockedStatus)
-            and self.unit.status.message != CORRUPT_PROMETHEUS_CONFIG_MESSAGE
-        ):
-            return
+            except ChangeError as e:
+                logger.error(
+                    "Failed to replan; pebble plan: %s; %s", container.get_plan().to_dict(), str(e)
+                )
+                self.unit.status = BlockedStatus(REPLAN_FAILED_MESSAGE)
+                return
 
-        self.unit.status = ActiveStatus()
+        # Auto-heal only from statuses managed by this method
+        if self.unit.status in [
+            WaitingStatus("Waiting for Pebble ready"),
+            BlockedStatus(CORRUPT_PROMETHEUS_CONFIG_MESSAGE),
+            BlockedStatus(REPLAN_FAILED_MESSAGE),
+        ]:
+            self.unit.status = ActiveStatus()
```

Comment:
I'm not sure about this approach for auto-heal:
- we probably want to heal only from statuses set by this method
- this exact approach would not work for variable messages e.g. `BlockedStatus(f"Invalid time spec : {timeval}")`
- for some reason `begin_with_initial_hooks` ends up with `Maintenance` (which the existing code masks) -- could this be a harness bug?

Both approaches -- touch status only in _configure, or auto-heal only from own statuses -- seem inadequate.

@balbirthomas @simskij

Reply:
With this new change of mine (only heal from own statuses) the itests fail too because they're [stuck on Maintenance or Unknown status](https://github.com/canonical/prometheus-k8s-operator/runs/5576228211?check_suite_focus=true#step:4:1449).

This is because Juju doesn't change it to Active on its own:

```shell
mkdir blank
charmcraft init -p blank
cd blank
# comment out: self.unit.status = ActiveStatus()
charmcraft pack
juju deploy ./blank_ubuntu-20.04-amd64.charm --resource httpbin-image=kennethreitz/httpbin
```

and the charm will go into Maintenance and at the end of startup change to unknown:

```python
# https://github.com/canonical/operator/blob/45e1e5a9e4e99492d09423436b8378f955c4f0a3/ops/model.py#L885
class UnknownStatus(StatusBase):
    """The unit status is unknown.
    A unit-agent has finished calling install, config-changed and start, but the
    charm has not called status-set yet.
    """
```

So I added Maintenance to the self healing list and now tests pass. Two questions:
1. Do we want this pattern?
2. Why some itests were stuck on Maintenance while others on Unknown?

Reply:
(2) suggests that the units are doing different things. Do you see a difference between the status of the leader unit and that of the followers maybe?